herddb-core/src/main/java/herddb/core/TableManager.java

```diff
@@ -2120,26 +2150,44 @@ private TableCheckpoint checkpoint(double dirtyThreshold, double fillThreshold,
                 flushedPages.add(weighted.pageId);
                 ++flushedSmallPages;
 
-                final DataPage dataPage = pages.get(weighted.pageId);
+                final boolean currentPageWasInMemory;
+                final DataPage dataPage = pages.remove(weighted.pageId);
                 final Collection<Record> records;
                 if (dataPage == null) {
                     records = dataStorageManager.readPage(tableSpaceUUID, table.uuid, weighted.pageId);
+                    currentPageWasInMemory = false;
                     LOGGER.log(Level.FINEST, "loaded small page {0} on tmp buffer: {1} records", new Object[]{weighted.pageId, records.size()});
                 } else {
                     records = dataPage.data.values();
+
+                    /* The page was found in memory. Currently built page should go into memory */
+                    currentPageWasInMemory = true;
+
+                    /* Current dirty record page isn't known to page replacement policy */
+                    if (currentDirtyRecordsPage.get() != dataPage.pageId) {
+                        pageReplacementPolicy.remove(dataPage);
+                    }
                 }
 
                 for (Record record : records) {
                     /* Flush the page if it would exceed max page size */
                     final long recordSize = DataPage.estimateEntrySize(record);
                     if (bufferPageSize + recordSize > maxLogicalPageSize) {
-                        createImmutablePage(buffer, bufferPageSize);
+                        /* Keep in memory? Actually yes */
```

Comment:
this comment is not clear.
if you are sure that keepFlushedPageInMemory is always true to be it is better to:
- add an assertion
- use the constant 'true' /* keepFlushedPageInMemory */ in the line 2177 below

Reply:
Is a dev leftover... It should be /* Do no keep in memory old untouched records */ as line 2105
I'll change it

Reply:
Log fixed